skada/deep/losses.py

```diff
@@ -189,6 +192,142 @@ def dan_loss(features_s, features_t, sigmas=None):
     return loss
 
 
+def cdd_loss(
+    y_s,
+    features_s,
+    features_t,
+    source_centroids=None,
+    sigmas=None,
+    distance_threshold=0.5,
+    class_threshold=3,
+):
+    """Define the contrastive domain discrepancy loss based on [33]_.
+
+    Parameters
+    ----------
+    y_s : tensor
+        labels of the source data used to compute the loss.
+    features_s : tensor
+        features of the source data used to compute the loss.
+    features_t : tensor
+        features of the target data used to compute the loss.
+    source_centroids : tensor
+        Pre-computed source centroids.
+    sigmas : array like, default=None,
+        If array, sigmas used for the multi gaussian kernel.
+        If None, uses sigmas proposed  in [1]_.
+    distance_threshold : float, optional (default=0.5)
+        Distance threshold for discarding the samples that are
+        to far from the centroids.
+    class_threshold : int, optional (default=3)
+        Minimum number of samples in a class to be considered for the loss.
+
+    Returns
+    -------
+    loss : float
+        The loss of the method.
+
+    References
+    ----------
+    .. [33] Kang, G., Jiang, L., Yang, Y., & Hauptmann, A. G. (2019).
+           Contrastive adaptation network for unsupervised domain adaptation.
+           In Proceedings of the IEEE/CVF Conference on Computer Vision
+           and Pattern Recognition (pp. 4893-4902).
+    """
+    n_classes = len(y_s.unique())
+
+    # Use pre-computed source centroids
+    if source_centroids is None:
+        warnings.warn(
+            "Source centroids are not computed for the whole training set, "
+            "computing them on the current batch set."
+        )
+
+        source_centroids = []
+
+        for c in range(n_classes):
+            mask = y_s == c
+            if mask.sum() > 0:
+                class_features = features_s[mask]
+                normalized_features = F.normalize(class_features, p=2, dim=1)
+                centroid = normalized_features.mean(dim=0)
+                source_centroids.append(centroid)
+
+    # Use source centroids to initialize target clustering
+    target_kmeans = SphericalKMeans(
```

Comment:
It should not be done also at the beginning of the epochs?

Reply:
you're right